File: typescript/src/0979-distributeCoins.ts
import TreeNode from './treeNode';

function distributeCoins(root: TreeNode | null): number {
  let result = 0;
  function checkNode(node: TreeNode | null): number {
    if (!node) return 0;
    const tmp1 = checkNode(node.left);
    const tmp2 = checkNode(node.right);
    result = result + Math.abs(tmp1) + Math.abs(tmp2);
    return node.val + tmp1 + tmp2 - 1;
  }
  checkNode(root);
  return result;
}

export {distributeCoins};

File: python/src/0979-distributeCoins.py
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        self.result = 0
        def checkNode(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            tmp1 = checkNode(node.left)
            tmp2 = checkNode(node.right)
            self.result = self.result + abs(tmp1) + abs(tmp2)
            return node.val + tmp1 + tmp2 - 1
        checkNode(root)
        return self.result
